ru.bmstu.rk9.rao.ui/src/ru/bmstu/rk9/rao/ui/serialization/SerializationConfigView.java

```diff
@@ -100,163 +168,307 @@ public void doubleClick(DoubleClickEvent event) {
-						return;
-
-					IEditorPart editor = partRef.getPage().getActiveEditor();
-					IXtextDocument document = ((XtextEditor) editor)
-							.getDocument();
-
-					EList<EObject> contents = document.readOnly(
-							new IUnitOfWork<XtextResource, XtextResource>() {
-								public XtextResource exec(XtextResource state) {
-									return state;
-								}
-							}).getContents();
-
-					if (contents.isEmpty())
-						return;
-
-					RaoModel model = (RaoModel) contents.get(0);
-					SerializationNode newModel = addModel(model.eResource());
-					modelNodes.put(partRef, newModel);
-
-					List<SerializationNode> modelsWithSameName = serializationConfig
-							.findModelsWithSameName(newModel.getFullName());
-					if (modelsWithSameName.size() > 1)
-						for (SerializationNode node : modelsWithSameName)
-							node.mustShowFullName(true);
-
-					if (serializationTreeViewer.getInput() == null)
-						serializationTreeViewer.setInput(serializationConfig);
-
-					document.addModelListener(new IXtextModelListener() {
+	IResourceChangeListener resourceChangeListener = new IResourceChangeListener() {
+		@Override
+		public void resourceChanged(IResourceChangeEvent event) {
+			switch (event.getType()) {
+			case IResourceChangeEvent.POST_CHANGE:
+				try {
+					event.getDelta().accept(new IResourceDeltaVisitor() {
 						@Override
-						public void modelChanged(XtextResource resource) {
-							updateInput(resource);
+						public boolean visit(IResourceDelta delta)
+								throws CoreException {
+							IResource resource = delta.getResource();
+							switch (resource.getType()) {
+							case IResource.ROOT:
+							case IResource.PROJECT:
+							case IResource.FOLDER:
+								return true;
+							case IResource.FILE:
+								handleFileChanges(delta);
+								break;
+							}
+							return false;
+						}
+
+						private final void handleFileChanges(IResourceDelta delta) {
+							if (!(delta.getResource() instanceof IFile))
+								return;
+							switch (delta.getKind()) {
+							case IResourceDelta.ADDED: {
+								IProject project = delta.getResource()
+										.getProject();
+								if (!project.isAccessible())
+									return;
+
+								updateProjectContents(project);
+								break;
+							}
+							case IResourceDelta.CHANGED: {
+								IFile file = (IFile) delta.getResource();
+								if (!file.isAccessible())
+									return;
+
+								if (!("rao".equalsIgnoreCase(file
+										.getFileExtension())))
+									return;
+
+								updateFileContents(file);
+								break;
+							}
+							case IResourceDelta.REMOVED: {
+								IFile file = (IFile) delta.getResource();
+								if (!file.isAccessible())
+									return;
+
+								if (!("rao".equalsIgnoreCase(file
+										.getFileExtension())))
+									return;
+
+								removeFileFromTree(file);
+								break;
+							}
+							default:
+								break;
+							}
 						}
 					});
+				} catch (CoreException e) {
+					e.printStackTrace();
```

Comment:
http://ru.stackoverflow.com/questions/332814/%D0%9B%D0%BE%D0%B3%D0%B8%D1%80%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5-%D0%B2-java